icu4c/source/common/cjmlbe.cpp

```diff
@@ -0,0 +1,462 @@
+        featureList->adoptElement(concatUnicodeString("TB1:", UnicodeString(b1).append(b2).append(b3), status), status);
+    }
+    if (b2 != UnicodeString(INVALID) && b3 != UnicodeString(INVALID) && b4 != UnicodeString(INVALID)) {
+        featureList->adoptElement(concatUnicodeString("TB2:", UnicodeString(b2).append(b3).append(b4), status), status);
+    }
+    if (b3 != UnicodeString(INVALID) && b4 != UnicodeString(INVALID) && b5 != UnicodeString(INVALID)) {
+        featureList->adoptElement(concatUnicodeString("TB3:", UnicodeString(b3).append(b4).append(b5), status), status);
+    }
+    if (b4 != UnicodeString(INVALID) && b5 != UnicodeString(INVALID) && b6 != UnicodeString(INVALID)) {
+        featureList->adoptElement(concatUnicodeString("TB4:", UnicodeString(b4).append(b5).append(b6), status), status);
+    }
+
+    int32_t score = negativeSum;
+    for (int32_t j = 0; j < featureList->size(); j++) {
+        UnicodeString* key = (UnicodeString*)featureList->elementAt(j);
+        if(fModel.containsKey(*key)) {
+            score += (2 * fModel.geti(*key));
+        }
+    }
+
+    if (score > 0) {
+        boundary.addElement(index, status);
+        numBreaks++;
+    }
+
+    featureList->removeAllElements();
+    delete featureList;
+}
+
+void CjMLBreakEngine::initElementList(UnicodeString& inString, UVector* elementList, UErrorCode& status) const {
+    UChar32 w1, w2, w3, w4, w5, w6;
+    UnicodeString b1, b2, b3, b4, b5, b6;
+    w1 = INVALID;
+    w2 = INVALID;
+    w3 = inString.char32At(0);
+    w4 = inString.char32At(1);
+    w5 = 2 < inString.length() ? inString.char32At(2) : INVALID;
+    w6 = 3 < inString.length() ? inString.char32At(3) : INVALID;
+
+    b1 = UnicodeString(INVALID);
+    b2 = UnicodeString(INVALID);
+    b3 = getUnicodeBlockIndex(w3, status);
+    b4 = getUnicodeBlockIndex(w4, status);
+    b5 = getUnicodeBlockIndex(w5, status);
+    b6 = getUnicodeBlockIndex(w6, status);
+
+    elementList->addElement(new Element(w1, b1), status);
+    elementList->addElement(new Element(w2, b2), status);
+    elementList->addElement(new Element(w3, b3), status);
+    elementList->addElement(new Element(w4, b4), status);
+    elementList->addElement(new Element(w5, b5), status);
+    elementList->addElement(new Element(w6, b6), status);
+}
+
+/**
+ * Convert an integer, positive or negative, to a character string radix 10.
+ */
+static char* itoa(int32_t i, char* buf) {
+    char* result = buf;
+    // Handle negative
+    if (i < 0) {
+        *buf++ = '-';
+        i = -i;
+    }
+    // Output digits in reverse order
+    char* p = buf;
+    do {
+        *p++ = (char)('0' + (i % 10));
+        i /= 10;
+    } while (i);
+    *p-- = 0;
+    // Reverse the string
+    while (buf < p) {
+        char c = *buf;
+        *buf++ = *p;
+        *p-- = c;
+    }
+
+    return result;
+}
+
+UnicodeString CjMLBreakEngine::getUnicodeBlockIndex(UChar32 ch, UErrorCode& status) const{
```

Comment:
Why are we mapping a character to a 3-digit string for its Unicode block index?
- There are only a few hundred blocks. We could easily just return a single unique code unit per block.
- The block indexes defined by this binary search will change with nearly every version of Unicode. Will we have to re-train the model for each version?
  - For example, this data is based on Unicode 14, but Unicode 15 will be released tomorrow (sep13) with several new blocks, and ICU 72 will ship with that.

Reply:
> The block indexes defined by this binary search will change with nearly every version of Unicode. Will we have to re-train the model for each version?

Good to know. BudouX uses Unicode block to learn more generalized rules than character-based ones (e.g. breaking after the "CJK Symbols and Punctuation" block vs breaking after the "。" character), which contributes to minimize the file size of the model.
That said, BudouX uses a limited number of Unicode blocks, so it should be fine to turn them into a unique code unit per block instead of a 3-digit string.

Reply:
FYI The Japanese BudouX model only uses 7 Unicode blocks below.
- General Punctuation
- Miscellaneous Symbols
- CJK Symbols and Punctuation
- Hiragana
- Katakana
- CJK Unified Ideographs
- Halfwidth and Fullwidth Forms

I think we can prepare unique codes for these blocks and replace the 3-digit strings with them.

Reply:
Done. Change to use these seven UBlockCode.

Reply:
Thanks, I will look at the code changes.

However, @tushuhei this discussion makes me wonder, because Unicode blocks are just a way to manage the encoding of characters (and some default property values). They are not very useful for properties. For example, if your model has an input value for the "CJK Unified Ideographs" block, then what about the 8 blocks "CJK Extension A" to "CJK Extension H"? They contain the same type of characters. The Script property seems much more reliable for this kind of thing: sc=Hani vs. Kana vs. Hira etc.

Similar for "General Punctuation" and "Miscellaneous Symbols". There are dozens of blocks with some or many punctuation and symbol characters. Should punctuation and symbols be treated consistently regardless of whether they happen to be encoded in these two blocks in your list? Of course, in a stand-alone BudouX project, adding character properties data could increase the data size noticeably, but in ICU we already have all of these properties available.

I suspect that this is a discussion that will go beyond this pull request -- that is, not blocking. Happy to meet and discuss at some point.